### PR TITLE
Tag API v1

### LINKED ISSUE #263: Biome tags
allow people to add tags for biomes. Especially useful for modding, where the primary way to add features to existing biomes is by looping over them and finding what you want. This would allow us to loop over it and check if the tag contains the biome.

### PR DESCRIPTION
**Pull request by i509VCB**
This PR deprecates the old V0 `tag-extensions` module and replaces it with `fabric-tag-api-v1`

~~This resolves #263~~
Extra tag types sadly impose a client and server mod requirement which is unacceptable for vanilla clients. I will open up my own libraries or future pull requests to handle that special case later.

### CODE REVIEW

File: fabric-tag-extensions-v1/src/main/java/net/fabricmc/fabric/api/tag/extension/v1/FabricTag.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.tag.extension.v1;
+
+/**
+ * Interface implemented by {@link net.minecraft.tag.Tag} instances when
+ * Fabric API is present.
+ *
+ * @param <T>
```
Comment (Juuxel) [inline fabric-tag-extensions-v1/src/main/java/net/fabricmc/fabric/api/tag/extension/v1/FabricTag.java]:
```suggestion
 * @param <T> the type contained in the tag
```

File: fabric-tag-extensions-v1/src/main/java/net/fabricmc/fabric/api/tag/extension/v1/FabricTagBuilder.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.tag.extension.v1;
+
+/**
+ * Interface implemented by {@link net.minecraft.tag.Tag.Builder} instances when
+ * Fabric API is present.
+ *
+ * @param <T>
```
Comment (Juuxel) [inline fabric-tag-extensions-v1/src/main/java/net/fabricmc/fabric/api/tag/extension/v1/FabricTagBuilder.java]:
```suggestion
 * @param <T> the type contained in the tag
```

File: fabric-tag-extensions-v1/src/main/java/net/fabricmc/fabric/api/tag/extension/v1/TagRegistry.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+		return create(id, BiomeTags::getContainer);
+	}
+
+	public static Tag<DimensionType> dimension(Identifier id) {
+		return create(id, DimensionTags::getContainer);
+	}
+
+	private static Field fluidTagContainer;
+
+	private static TagContainer<Fluid> getFluidTagContainer() {
```
Comment (Juuxel) [inline fabric-tag-extensions-v1/src/main/java/net/fabricmc/fabric/api/tag/extension/v1/TagRegistry.java]: I think this can use a static `@Accessor`.

**Review by Juuxel — CHANGES_REQUESTED** [review]

File: fabric-tag-extensions-v1/src/main/java/net/fabricmc/fabric/impl/tag/extension/v1/RegistryTagReloadListener.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+import net.minecraft.tag.RegistryTagContainer;
+import net.minecraft.tag.Tag;
+import net.minecraft.tag.TagContainer;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.registry.Registry;
+
+import net.fabricmc.fabric.api.resource.SimpleSynchronousResourceReloadListener;
+import net.fabricmc.fabric.api.tag.extension.v1.TagRegistry;
+
+public class RegistryTagReloadListener implements SimpleSynchronousResourceReloadListener {
```
Comment (liach) [inline fabric-tag-extensions-v1/src/main/java/net/fabricmc/fabric/impl/tag/extension/v1/RegistryTagReloadListener.java]: Please make this async.

**Review by liach — CHANGES_REQUESTED** [review]

**Comment by liach** [comment]
i think you made the api's scope too big. the old api is like plugin usable and now you add the networking, which really needs a client

**Comment by i509VCB** [comment]
I could split the registry tags into a separate module if we feel the scope is now too big.

The reason I added the tag synchronization was to consider the fact that some mods on the client may want to sort enchantments by category or use biome tags.

EDIT: Obviously it's a draft just for that reason since the scope is still being decided.

**Comment by liach** [comment]
after another look i think this setup is fine. just make sure you only send the packet if connection is not local and client can actually receive this packet type

**Comment by i509VCB** [comment]
I need to rebase this

File: build.gradle [hunk 5]
```diff
@@ -225,7 +235,7 @@ curseforge {
 		id = '306612'
 		changelog = 'A changelog can be found at https://github.com/FabricMC/fabric/commits/master'
 		releaseType = 'beta'
-		addGameVersion '1.15-Snapshot'
+		addGameVersion '1.15.2'
```
Comment (liach) [inline build.gradle]: Why? this appears wrong to me. Maybe you've messed up git

**Review by liach — CHANGES_REQUESTED** [review]

File: build.gradle (hunk 5 above)
Comment (i509VCB) [reply build.gradle]: something screwed up and I need to go look into why it messed up (probably cherrypick it to fix)

Comment (i509VCB) [reply build.gradle]: Hmm due to the tag changes, should I count on doing this in 1.15 or 1.16.

I know modmuss had a blast backporting the object builders when I did them for 1.16 instead.

Comment (liach) [reply build.gradle]: You probably shouldn't change the base branch of pr to 1.16; should change it back to 1.15, and cherry pick the squashed commit onto 1.16 and finally reset the head of `feature/registry-tags` to the repicked commit, then change the pr's base branch to 1.16

**Comment by i509VCB** [comment]
I am reconsidering some design aspects of this PR. 1.16 having the changes with dimension types does make this a bit of fun.

**Comment by liach** [comment]
The tagging shouldn't hurt too much here. I expect dimension tags to work like function tags instead of other registry tags.

**Comment by i509VCB** [comment]
I decided to bind dimension tags to the current dimension tracker, and the registry it holds. Locally I am working on it and I've decided so far to support the following:

```
Biomes
BlockEntityTypes
DimensionTypes
Enchantments
Painting(Motives/Motifs/Whatever)
ParticleTypes
PointOf InterestTypes
SoundEvents
StatusEffects
VillagerProfessions
VillagerTypes.
```

After some further realization, this will be waiting till modmuss finishes and commits his registry sync fixes, since I have to mark all those registries as `SYNCED` in order to have tags sync properly (since they use the raw ids in packets).

**Comment by liach** [comment]
Why must you sync tags? Imo tag syncing should be optional (vanilla does not sync function tags for instance) and if a registry is int-id synced, we can sync tag elements by int id safely. No need to touch registry at all.

**Comment by i509VCB** [comment]
I think the expectation is that these types of tags are synced. Function tags are the exception in this case since they are not bound to a registry.

On the case of optional or not, I only require the client to understand the packet in order to sync the custom tags. So a vanilla client would not have custom tags synced. ClientWorld does call some BlockTags checks, so I'd assume these tags should be synced to the client.

> if a registry is int-id synced, we can sync tag elements by int id safely. No need to touch registry at all.

To avoid syncing ids would require some custom packet logic to use identifiers instead. It's possible but I could see scenarios where we hit the maximum packet payload easily. (Even if we sent each type of tag container individually).

**Comment by i509VCB** [comment]
Finally put my changes on branch. This is now targetting 1.16.

Note this is being cleaned up a bit before the final review cycle comes

**Comment by i509VCB** [comment]
Okay so I did some experimenting with dimension types. The require changes are too invasive for fabric api.

This includes moving where RegistryTracker is initialized in `server.Main`, Replacing all calls to `ServerResourceManager.reload()` to a new call which does the same thing with a registry tracker.

So the current tagged types are final as they are.

Since Dimension tags are possible but not done easily right now, I will make a mod to add taggable dimensions and use that as a better gauge of whether it belongs in fabric api at a later date.

Will write up some final testing for this tomorrow and then I can open this up to review.

File: fabric-extra-tag-api-v1/build.gradle [hunk 6]
```diff
@@ -0,0 +1,13 @@
+archivesBaseName = "fabric-registry-tag-api-v1"
```
Comment (Prospector) [inline fabric-extra-tag-api-v1/build.gradle]: Is it `registry-tag-api` or `extra-tag-api`? I'll be honest, I'm really not a fan of `extra-tag-api` but unfortunately I don't really have an alternative idea. Maybe `fabric-more-tag-types-api`?

File: fabric-tag-api-v1/src/main/java/net/fabricmc/fabric/api/tag/v1/TagRegistry.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+/**
+ * Helper methods for registering Tags.
+ */
+public final class TagRegistry {
+	private TagRegistry() {
+	}
+
+	public static Tag.Identified<Block> block(Identifier id) {
+		return getBlockAccessor().get(id.toString());
```
Comment (Prospector) [inline fabric-tag-api-v1/src/main/java/net/fabricmc/fabric/api/tag/v1/TagRegistry.java]: Why not just use an Invoker to invoke vanilla's `register` methods directly?

Comment (i509VCB) [reply fabric-tag-api-v1/src/main/java/net/fabricmc/fabric/api/tag/v1/TagRegistry.java]: both have the same cost anyways. From experience with registries, we should not use the default register methods due to defaulting to minecraft namespace.

File: fabric-extra-tag-api-v1/build.gradle (hunk 6 above)
Comment (i509VCB) [reply fabric-extra-tag-api-v1/build.gradle]: Tbf `registry-tag-api` was the original name, but the fact we don't cover every registry makes it misleading.

The name could be improved for sure.

**Comment by i509VCB** [comment]
I need to update the module to use identified tags for initial release. Extra tag module at a later date probably.

**Comment by i509VCB** [comment]
Rebased to latest.

**Review by Juuxel — CHANGES_REQUESTED** [review]
This looks really nice. Just one note: `tag-extensions-v0`'s version should be bumped.

(Also, couldn't you just use an `@Invoker` for the `BlockTags.register` method etc instead of accessing the `RequiredTagList`?)

**Comment by i509VCB** [comment]
Are you referring to the `required` or regular method? These methods both do different things.

**Comment by Juuxel** [comment]
The `required` method. At least `BlockTags.register` registers the tags as required tags (found this out when porting Adorn to use the vanilla system as I needed `Tag.Identified`).

**Comment by i509VCB** [comment]
Moved to invoker on register, it is functionally the same as field access

**Pull request closed without merging**